### ISSUE TITLE
packages installed with `pip install --user` are visible inside buildpack process, resulting in packages skipped in artifact

### ISSUE THREAD

**Issue by cxmcc** [heroku/heroku-buildpack-python#648]
### Summary
- If we run `pip install --user <package>`, the package will be installed in `~/.local/lib`
- `~/.local/lib` is still visible to buildpack, buildpack will skip installing those packages.

### Reproducing steps
```
# installing a package locally
$ pip install --user boto==2.48.0 > /dev/null

$ pip freeze --user boto==2.48.0
boto==2.48.0

# requirements.txt contains the above pacakge
$ cat > $BUILD_DIR/requirements.txt <<EOF
boto==2.48.0
boto3
EOF

$ bin/compile "$BUILD_DIR" "$CACHE_DIR"
# boto will NOT be installed, boto3 will be installed...
```

### Potential fixes
- Print a warning if `pip freeze --user` returns something
- Change `$HOME` to something buildpack controls
- Maybe there are some options in pip that can be used to ignore `~/.local/*`

**Comment by kennethreitz** [comment]
sounds like a problem with your platform, not with this buildpack.

**Comment by cxmcc** [comment]
@kennethreitz Yes and No.
- It's quite unlikely that a platform will run `pip install --user` to install some packages locally instead of globally. Somehow we did. However, it's arguably a more secure practice.
- We use buildpack with the assumption that python packages installed outside of buildpack environment will not affect the functionality of buildpack (packages installed with `pip install` as root, and packages installed with `pip install --user` in my case). But that's not the case here.
- Buildpack silently fails because environment/platform has an undocumented but reasonable edge case. A warning message would help prevent wasting troubleshooting time for future users with the same situation.
I may miss something, if so please let me know. thanks.

**Comment by edmorley** [comment]
It's possible others picture this buildpack (or buildpacks in general) differently, but I'd always imagined them to only be run inside a disposable environment (eg docker image, VM, ...) and not directly on the host machine. (For example this buildpack creates symlinks at `/app` so isn't really something I'd want to run directly on my machine; I'm sure the same is similar of other buildpacks.)

In a disposable/scripted environment, for what reasons would someone run `pip install --user` manually? Or is the buildpack being used directly on the host after all?

**Comment by cxmcc** [comment]
- We use a pool of jenkins slave nodes to run mostly buildpacks, along with some other tasks.
- The hosts are disposable but do not use another layer of visualization/container (hey, good question actually, why don't we just use containers all the way instead of buildpack).
- We do not immediately terminate the hosts once buildpack is done to avoid the overhead of provisioning.
- The `pip install` is scripted by another task.

I hope this explains what has happened. "disposable" is not enough. It needs to be truly immutable to prevent such problems.

Yeah. I agree that visualization solves quite some problems. I noticed the dockerfile here, which did not exist when I started using buildpack. Maybe we should use that.

Please close Issue if determined to be wont-fix. It's reasonable to me if buildpack can only run in container/vm.

**Comment by kennethreitz** [comment]
i think this is out of scope.
